pkg/work/spoke/conditions/reader.go

```diff
@@ -0,0 +1,281 @@
+) (metav1.Condition, int64, error) {
+	switch rule.Type {
+	case workapiv1.WellKnownConditionsType:
+		r := s.wellKnownConditions.GetRuleByKindCondition(obj.GroupVersionKind(), rule.Condition)
+		if len(r.CelExpressions) == 0 {
+			err := fmt.Errorf("cannot find the wellknown conditions for resource with gvk %s", obj.GroupVersionKind().String())
+			return metav1.Condition{
+				Type:    rule.Condition,
+				Status:  metav1.ConditionUnknown,
+				Reason:  workapiv1.ConditionRuleInvalid,
+				Message: err.Error(),
+			}, budget, err
+		}
+
+		// Check for supported overrides in rule
+		if rule.Message != "" {
+			r.Message = rule.Message
+		}
+		if rule.MessageExpression != "" {
+			r.MessageExpression = rule.MessageExpression
+		}
+		return s.getConditionByCelRule(ctx, obj, r, budget)
+	case workapiv1.CelConditionExpressionsType:
+		return s.getConditionByCelRule(ctx, obj, rule, budget)
+	default:
+		err := fmt.Errorf("unrecognized condition rule type %s", rule.Type)
+		return metav1.Condition{
+			Type:    rule.Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  workapiv1.ConditionRuleInvalid,
+			Message: err.Error(),
+		}, budget, err
+	}
+}
+
+func (s *ConditionReader) getConditionByCelRule(
+	ctx context.Context, obj *unstructured.Unstructured, rule workapiv1.ConditionRule, budget int64,
+) (metav1.Condition, int64, error) {
+	var message string
+	var err error
+	status, reason, remainingBudget, err := s.evaluateCelExpressions(ctx, obj, rule.CelExpressions, budget)
+	if err != nil {
+		message = err.Error()
+	} else {
+		message, remainingBudget, err = s.getConditionMessageByRule(ctx, obj, rule, status == metav1.ConditionTrue, remainingBudget)
+		if err != nil && message == "" {
+			message = err.Error()
+		}
+	}
+
+	return metav1.Condition{
+		Type:    rule.Condition,
+		Status:  status,
+		Message: message,
+		Reason:  reason,
+	}, remainingBudget, err
+}
+
+func (s *ConditionReader) evaluateCelExpressions(
+	ctx context.Context, obj *unstructured.Unstructured, expressions []string, budget int64,
+) (status metav1.ConditionStatus, reason string, remainingBudget int64, err error) {
+	remainingBudget = budget
+	estimator := newEstimator()
+	for _, expression := range expressions {
+		ast, iss := s.ruleEnv.Compile(expression)
+		err = iss.Err()
+		if err != nil {
+			return metav1.ConditionFalse, workapiv1.ConditionRuleExpressionError, remainingBudget, err
+		}
+
+		prg, err := s.ruleEnv.Program(
+			ast,
+			cel.CostLimit(celconfig.PerCallLimit),
+			cel.CostTracking(estimator),
+			cel.InterruptCheckFrequency(celconfig.CheckFrequency),
+		)
+		if err != nil {
+			return metav1.ConditionUnknown, workapiv1.ConditionRuleInternalError, remainingBudget, err
+		}
+
+		out, newBudget, err := evaluate(ctx, prg, remainingBudget, expression, map[string]any{
+			"object": obj.Object,
+		})
+		if err != nil {
+			return metav1.ConditionFalse, workapiv1.ConditionRuleExpressionError, remainingBudget, err
```

Comment:
could you add some comment on why here is false and line156 return unknown?

Reply:
It's a bit arbitrary. I returned unknown on line 156 because creating the program is not in any way controlled by the user that wrote the CEL expression. Lines 146 and 163 return errors with condition false because they relate to errors in the CEL expression itself, so the user is in control. In other words, I chose to represent a bad rule as false rather than unknown and reserve unknown for cases where the user has no control.